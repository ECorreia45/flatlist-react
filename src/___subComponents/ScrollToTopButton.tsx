import {element, func, node, number, oneOf, oneOfType} from 'prop-types';
import React, {createRef, Ref, useEffect} from 'react';
import {isFunction} from '../___utils/isType';
import {btnPosition} from './uiFunctions';

const ScrollToTopButton = (props: any) => {
    const anchor: Ref<HTMLElement> = createRef();
    const {button, scrollToTopPosition, scrollToTopPadding, scrollToTopOffset} = props;
    const btn = isFunction(button) ? button() : button;

    // eslint-disable-next-line consistent-return
    useEffect(() => {
        const buttonElement = (anchor as any).current.nextElementSibling;
        const container = (anchor as any).current.parentNode;
        (anchor as any).current.remove();

        const positionBtn = btnPosition(container, buttonElement);
        const pos = scrollToTopPosition.split(' ');
        const updateBtnPosition = () => positionBtn(pos[0], pos[1], scrollToTopPadding, scrollToTopOffset);
        window.addEventListener('resize', updateBtnPosition);
        container.addEventListener('scroll', updateBtnPosition);

        buttonElement.addEventListener('click', () => {
            container.scrollTo({
                top: 0,
                behavior: 'smooth'
            });
        });

        setTimeout(() => updateBtnPosition(), 250);

        return () => {
            window.removeEventListener('resize', updateBtnPosition);
        };
    }, []);

    return (
        <>
            <span ref={anchor} style={{display: 'none'}}/>
            {button ? btn : <button type="button">To Top</button>}
        </>
    );
};

ScrollToTopButton.propTypes = {
    button: oneOfType([node, element, func]),
    scrollToTopPosition: oneOf(['top right', 'top left', 'bottom right', 'bottom left']),
    scrollToTopPadding: number,
    scrollToTopOffset: number
};

ScrollToTopButton.defaultProps = {
    button: null,
    scrollToTopPadding: 20,
    scrollToTopOffset: 50,
    scrollToTopPosition: 'bottom right'
};

export default ScrollToTopButton;
