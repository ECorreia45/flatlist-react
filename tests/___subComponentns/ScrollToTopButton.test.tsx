import {render} from '@testing-library/react';
import React, {createRef} from 'react';
import ScrollToTopButton from '../../src/___subComponents/ScrollToTopButton';

describe('ScrollToTopButton', () => {
    it('Should match snapshot', () => {
        const ref: any = createRef();
        const {asFragment} = render(
            <div ref={ref}>
                <ScrollToTopButton scrollingContainer={ref}/>
            </div>);

        expect(asFragment()).toMatchSnapshot();
    });

    it('Should have button with initial style', () => {
        const ref: any = createRef();
        const {getAllByText} = render(
            <div ref={ref}>
                <ScrollToTopButton scrollingContainer={ref}/>
            </div>
        );
        const btn = getAllByText('To Top')[0];

        expect(btn.style.cssText).toBe('position: absolute; z-index: 1; visibility: hidden;');
    });
});
