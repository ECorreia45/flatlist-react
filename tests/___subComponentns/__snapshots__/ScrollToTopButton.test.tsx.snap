// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollToTopButton Should match snapshot 1`] = `
<DocumentFragment>
  <button
    style="position: fixed; z-index: 1; visibility: hidden;"
    type="button"
  >
    To Top
  </button>
</DocumentFragment>
`;
